src/v/cloud_storage/remote_partition.cc

```diff
@@ -0,0 +1,652 @@
+    // performed.
+    auto delta = m.delta_offset == model::offset::min() ? model::offset(0)
+                                                        : m.delta_offset;
+    return m.base_offset - delta;
+}
+
+class record_batch_reader_impl final : public model::record_batch_reader::impl {
+    using remote_segment_list_t = std::vector<std::unique_ptr<remote_segment>>;
+    using remote_segment_iterator = remote_segment_list_t::iterator;
+
+public:
+    explicit record_batch_reader_impl(
+      const log_reader_config& config,
+      ss::lw_shared_ptr<remote_partition> part) noexcept
+      : _ctxlog(cst_log, _rtc, part->get_ntp().path())
+      , _partition(std::move(part))
+      , _it(_partition->_segments.begin()) {
+        if (config.abort_source) {
+            vlog(_ctxlog.debug, "abort_source is set");
+            auto sub = config.abort_source->get().subscribe([this]() noexcept {
+                vlog(_ctxlog.debug, "abort requested via config.abort_source");
+                _partition->evict_reader(std::move(_reader));
+                _it = _partition->_segments.end();
+            });
+            if (sub) {
+                _as_sub = std::move(*sub);
+            } else {
+                vlog(_ctxlog.debug, "abort_source is triggered in c-tor");
+                _it = _partition->_segments.end();
+                _reader = {};
+            }
+        }
+        if (!is_end_of_stream()) {
+            initialize_reader_state(config);
+        }
+    }
+
+    ~record_batch_reader_impl() override = default;
+    record_batch_reader_impl(record_batch_reader_impl&& o) noexcept = delete;
+    record_batch_reader_impl&
+    operator=(record_batch_reader_impl&& o) noexcept = delete;
+    record_batch_reader_impl(const record_batch_reader_impl& o) = delete;
+    record_batch_reader_impl& operator=(const record_batch_reader_impl& o)
+      = delete;
+
+    bool is_end_of_stream() const override {
+        return _it == _partition->_segments.end();
+    }
+
+    ss::future<storage_t>
+    do_load_slice(model::timeout_clock::time_point deadline) override {
+        try {
+            if (is_end_of_stream()) {
+                vlog(
+                  _ctxlog.debug,
+                  "record_batch_reader_impl do_load_slize - empty");
+                co_return storage_t{};
+            }
+            if (_reader->config().over_budget) {
+                vlog(_ctxlog.debug, "We're overbudget, stopping");
+                // We need to stop in such way that will keep the
+                // reader in the reusable state, so we could reuse
+                // it on next itertaion
+
+                // The existing state have to be rebuilt
+                _partition->return_reader(std::move(_reader), _it->second);
+                _it = _partition->_segments.end();
+                co_return storage_t{};
+            }
+            while (_reader) {
+                if (co_await maybe_reset_reader()) {
```

Comment:
Not sure but looks like while and if can be coalesced into a single loop `while (co_await maybe_reset_reader()) {...}`?